### DISCUSSION TITLE
[QUESTION] Old `engine-image-ei-*` not removed after upgrade

### DISCUSSION THREAD

**Discussion by aureq** [longhorn/longhorn#4522] [Q&A] [answered]
## Question

I just upgraded from `v1.2.3` to `v1.3.1` using `helm` as it's describe in the documentation. I noticed a bunch of pods being stopped and redeployed and after a few minutes everything appeared to be stable and there wasn't any crash loop.

Thereafter, I stopped all the workloads that were using a longhorn volume (using a combination of `helm` and `kubectl` as required). All of this went well.

However, I noticed the daemonSet `engine-image-ei-*` appears to be duplicated when I list them all

```
kubectl get daemonsets -n longhorn-system
NAME                       DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
engine-image-ei-766a591b   3         3         3       3            3           <none>          60m
engine-image-ei-fa2dfbf0   3         3         3       3            3           <none>          233d
longhorn-csi-plugin        3         3         3       3            3           <none>          59m
longhorn-manager           3         3         3       3            3           <none>          233d
```

If I describe `engine-image-ei-fa2dfbf0` and look for the container image, I obtain `longhornio/longhorn-engine:v1.2.3`.
```bash
kubectl get daemonsets -n longhorn-system engine-image-ei-fa2dfbf0 -o json | jq '.spec.template.spec.containers[].image'
```
However, running the same command on `engine-image-ei-766a591b` shows `longhornio/longhorn-engine:v1.3.1`

Is it expected that the 2 daemonSets are running simultaneously, and on 2 different version?

## Environment

 - Longhorn version: 1.3.1 (upgraded from 1.2.3)
 - Kubernetes version: v1.23.10+k3s1
 - Node config
   - OS type and version: Debian GNU/Linux 11 (bullseye)
   - CPU per node: 4c/8t
   - Memory per node: 32Gb
   - Disk type: nvme
   - Network bandwidth and latency between the nodes: 1Gpbs
 - Underlying Infrastructure: Baremetal
   - Number of nodes: 3

**Comment by derekbit** [comment]
The engine image is still used by instance-managers. Please refer to https://longhorn.io/kb/troubleshooting-some-old-instance-manager-pods-are-still-running-after-upgrade/.

IIRC, the engine image will be deleted after 60 minutes if there is no instance-manager using it.

**Comment by aureq** [comment]
Hi @derekbit thanks for the link. If I understand this correctly, somehow I have some PVC using the old engine which is why some `engine-image-ei-*` point to the old version. Is that correct?

However, it appears all my `instance-manager-(r|e)-*` have been upgraded to `v1_20220808`. Would it mean the new instance-manager is able to use the old `engine-image-ei-*` and the new one?

Using the CLI, is there a command to list the PVCs/PVs and the engine version being used? (I haven't enabled the UI)

Thanks

**Comment by mantissahz** [comment, marked as answer]
Hi @aureq,
You could use `kubeclt -n longhorn-system get volume [your volume name] -o yaml` and engine version is at field `engineImage` and `currentImage`.
You set the a value bigger than 0 to `concurrent-automatic-engine-upgrade-per-node-limit` by
`kubectl -n longhorn-system edit setting` to upgrade the engines of volumes automatically.
ref: https://longhorn.io/docs/1.3.1/deploy/upgrade/auto-upgrade-engine/

**Comment by derekbit** [comment]
> Hi @derekbit thanks for the link. If I understand this correctly, somehow I have some PVC using the old engine which is why some `engine-image-ei-*` point to the old version. Is that correct?

Correct.

> 
> However, it appears all my `instance-manager-(r|e)-*` have been upgraded to `v1_20220808`. Would it mean the new instance-manager is able to use the old `engine-image-ei-*` and the new one?

The newly created instance-managers should use the new engine image, but the old ones are still using the old image.
You can check if there is any instance-manager-r/e using the old image by
```
 kubectl -n longhorn-system get instancemanagers -o yaml | grep image
```

> 
> Using the CLI, is there a command to list the PVCs/PVs and the engine version being used? (I haven't enabled the UI)
> 
> Thanks

From the above check, you can find which instance-manager is using the old image. Then, from the instancemanager resource, you can futher find out the replicas and pvc.

**Comment by aureq** [comment]
Thank you to you both @derekbit @mantissahz 

I've been able to upgrade all my volumes without any issues. What I was missing was being able to edit the settings via the CLI. I think it would be most helpful if this command could be included as part of the official documentation (ie, not a GH discussion).

I'm also sharing a command that may be useful to others.

The command below lists, the longhorn volume name, the engineImage, the currentImage, the volume state, its robustness, the pvc namespace, the pvc name, the last known workload name.
```bash
kubectl get volumes -n longhorn-system  -o=custom-columns='NAME:.metadata.name,ENGINE_IMAGE:.spec.engineImage,CURRENT_IMAGE:.status.currentImage,STATE:.status.state,ROBUSTNESS:.status.robustness,PVC_NAMESPACE:.status.kubernetesStatus.namespace,PVC_NAME:.status.kubernetesStatus.pvcName,LAST_KNOWN_WORKLOAD_NAME:.status.kubernetesStatus.workloadsStatus[0].workloadName'
```

**Reply by innobead** [reply]
For editing settings via kubectl, please refer to https://longhorn.io/docs/1.3.1/advanced-resources/deploy/customizing-default-settings/#using-kubectl.

**Reply by aureq** [reply]
@innobead Great to see there's a doc for the CLI. That's awesome. 

Maybe as suggestions, this [page](https://longhorn.io/docs/1.3.1/deploy/upgrade/auto-upgrade-engine/) should point to the [page]( https://longhorn.io/docs/1.3.1/advanced-resources/deploy/customizing-default-settings/#using-kubectl) you mentioned so it's clear to reader the CLI is an option. I think what wasn't clear is `concurrent automatic engine upgrade per node limit` wasn't referred as both a UI and a CLI setting.
